Add PDF handling capabilities for ArXiv papers and update dependencies.

dimension-drifter opened the pull request:
## Description
Fixed reading of ArXiv pdf.

## Related Issues
Fixes #190 

## Notes
Have fixed it using some basic methods. It could be enhanced further with additional changes in a few more files and using a different approach all together.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - **Enhanced Academic Document Processing:** The application now automatically extracts and summarizes key details from scholarly PDFs, including counts of figures and tables.
  - **Improved Document Handling:** Upgraded image processing and optical character recognition provide a more robust analysis of academic content.
  - **Streamlined Task Management:** Automated detection of academic sources ensures efficient retrieval and processing of relevant document insights.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

giordano-lucas commented on pyproject.toml:
```diff
@@ -29,6 +29,9 @@ dependencies = [
     "python-dotenv>=1.0.1",
     "requests>=2.32.3",
     "tldextract>=5.2.0",
+    "pypdf2>=3.0.0",
+    "pillow>=10.0.0",
+    "pytesseract>=0.3.10",
```
this needs to be resolved

giordano-lucas commented:
@dimension-drifter can you edit the PR description to add more context to what you are trying to achieve with this PR exactly ? What's your use case ? Do you want to be able to scrape content from arxiv papers and convert that into an LLM ready format data ? I don't think the `handle_arxiv_task` really has its place within the core package of notte given how specific it is to your particular use case

dimension-drifter commented:
## Purpose
This PR enhances notte's capabilities to better evaluate AI agents' abilities to navigate and extract information from academic websites, particularly ArXiv. It adds specialized handling for ArXiv papers as part of the WebVoyager benchmark dataset.

## Context
The WebVoyager dataset already includes ArXiv-related tasks, but lacks efficient tools for handling PDF content and paper metadata extraction. Several tasks involve searching papers, accessing HTML versions, and extracting information from papers, which current implementations handle sub-optimally.

## Implementation Details
Added PDF handling utilities that can extract ArXiv IDs from URLs and access paper content.
Implemented figure and table counting capabilities for academic papers.
Created a specialized `handle_arxiv_task` method to streamline interactions with ArXiv's interface.

giordano-lucas commented on src/notte/agents/main.py:
```diff
@@ -56,3 +57,41 @@ async def async_run(self, task: str, url: str | None = None) -> AgentResponse:
         return asyncio.run(agent.run(task, url=url))
+
+    # Add to the relevant section that handles tasks
+    async def handle_arxiv_task(self, task: str, url: str) -> str:
```
`handle_arxiv_task` should be moved to `examples/axiv.py` (this is a arxiv specific use-case

giordano-lucas commented on src/notte/agents/main.py:
```diff
@@ -56,3 +57,41 @@ async def async_run(self, task: str, url: str | None = None) -> AgentResponse:
+            # Search for the paper
+            search_box = self.env.snapshot.dom_node.find("I1")  # Assuming I1 is the search box ID
+            if search_box:
+                await self.act(FillAction(id="I1", value=paper_title))
```
There is no `act` method on the `Agent` class, this cannot work. Have you tested it before actually creating the PR ?

giordano-lucas commented:
@dimension-drifter the example that you shared is actually not a agent, it is a simple script / web automation. You can simply use a NotteEnv as follows to perform your usecase.

```python

import asyncio
from notte_core.controller.actions import ClickAction, FillAction, GotoAction
from notte_browser.env import NotteEnv, NotteEnvConfig
import re



def get_arxviv_title(task: str, url: str) -> str | None:
    if "arxiv.org" in url and "figures" in task.lower() and "tables" in task.lower():
        # Extract paper title from task
        paper_title_match = re.search(r'"([^"]+)"', task)
        if not paper_title_match:
            return None
        paper_title = paper_title_match.group(1)
        return paper_title
    return None

def handle_pdf_content(url: str) -> tuple[int, int]:
    pass

async def handle_arxiv_task(task: str, url: str) -> str:
        """Handle ArXiv specific tasks."""
        paper_title = get_arxviv_title(task, url)
        if paper_title is None:
            return "Could not find the specified paper."

        async with NotteEnv(config=NotteEnvConfig()) as env:
            _ = await env.act(GotoAction(url=url))
            # Search for the paper
            search_box = env.snapshot.dom_node.find("I1")  # Assuming I1 is the search box ID
            if search_box:
                _ = await env.act(FillAction(id="I1", value=paper_title))
                _ = await env.act(ClickAction(id="B1"))  # Assuming B1 is the search button ID
            
            paper_id: str | None = None
            for node in env.snapshot.dom_node.flatten():
                if paper_title.lower() in (node.text or "").lower():
                    paper_id = node.id
                    break
            
            if paper_id:
                obs = await env.act(ClickAction(id=paper_id))
                
                # Now we're on the paper page, use PDF handler
                figures, tables = await handle_pdf_content(obs.metadata.url)
                return f"{figures} Figures, {tables} Tables."
            
            return "Could not find the specified paper."
        


if __name__ == "__main__":
    result = asyncio.run(handle_arxiv_task("Extract figures and tables from the paper", "https://arxiv.org/abs/2404.03977"))
    print(result)
```

giordano-lucas commented:
closing this PR since this is too specific to arxiv papers to be included in the core package.
